server/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java

```diff
@@ -489,8 +490,43 @@ public static GeoDistanceSortBuilder fromXContent(XContentParser parser, String
 
     @Override
     public SortFieldAndFormat build(QueryShardContext context) throws IOException {
+        GeoPoint[] localPoints = localPoints();
+        boolean reverse = order == SortOrder.DESC;
+        MultiValueMode localSortMode = localSortMode();
+        IndexGeoPointFieldData geoIndexFieldData = fieldData(context);
+        Nested nested = nested(context);
 
-        // validation was not available prior to 2.x, so to support bwc percolation queries we only ignore_malformed
+        if (geoIndexFieldData.getClass() == LatLonPointDVIndexFieldData.class // only works with 5.x geo_point
+                && nested == null
+                && localSortMode == MultiValueMode.MIN // LatLonDocValuesField internally picks the closest point
+                && unit == DistanceUnit.METERS
+                && reverse == false
+                && localPoints.length == 1) {
+            return new SortFieldAndFormat(
+                    LatLonDocValuesField.newDistanceSort(fieldName, localPoints[0].lat(), localPoints[0].lon()),
+                    DocValueFormat.RAW);
+        }
+
+        return new SortFieldAndFormat(
+                new SortField(fieldName, comparatorSource(localPoints, localSortMode, geoIndexFieldData, nested), reverse),
+                DocValueFormat.RAW);
+    }
+
+    @Override
+    public BucketedSort buildBucketedSort(QueryShardContext context) throws IOException {
+        GeoPoint[] localPoints = localPoints();
+        MultiValueMode localSortMode = localSortMode();
+        IndexGeoPointFieldData geoIndexFieldData = fieldData(context);
+        Nested nested = nested(context);
+
+        // TODO implement the single point optimization above 
```

Comment:
It looks like we actually have three ways to do geo distance - the optimization above and a single point to single point optimization inside `GeoUtils.distanceValues`. It isn't clear to me that it is worth doing a bunch of work to mimic the optimization above without benchmarks. Which I will write eventually.